MapleServer2/Packets/AchievePacket.cs

```diff
@@ -0,0 +1,81 @@
+using System.Collections.Generic;
+using System.Linq;
+using MaplePacketLib2.Tools;
+using MapleServer2.Constants;
+using MapleServer2.Types;
+
+namespace MapleServer2.Packets
+{
+    public static class AchievePacket
+    {
+        private enum AchievePacketMode : byte
+        {
+            TableStart = 0x0,
+            TableContent = 0x1,
+            Update = 0x2
+        }
+
+        private enum GradeStatus : byte
+        {
+            NotFinalGrade = 0x0,
+            FinalGrade = 0x3
+        }
+
+        public static Packet WriteTableStart()
+        {
+            PacketWriter pWriter = PacketWriter.Of(SendOp.ACHIEVE);
+            pWriter.WriteEnum(AchievePacketMode.TableStart);
+
+            return pWriter;
+        }
+
+        // packet from WriteTableStart() must be sent immediately before sending these packets
+        public static Packet WriteTableContent(List<Achieve> achieves)
+        {
+            PacketWriter pWriter = PacketWriter.Of(SendOp.ACHIEVE);
+            pWriter.WriteEnum(AchievePacketMode.TableContent);
+            pWriter.WriteInt(achieves.Count);
+
+            foreach (Achieve achieve in achieves)
+            {
+                pWriter.WriteInt(achieve.Id);
+                pWriter.WriteInt(1);            // unknown 'SS' ?
+                pWriter.Write(WriteIndividualAchieve(achieve).Buffer);
+            }
+
+            return pWriter;
+        }
+
+        public static Packet WriteUpdate(Achieve achieve)
+        {
+            PacketWriter pWriter = PacketWriter.Of(SendOp.ACHIEVE);
+            pWriter.WriteEnum(AchievePacketMode.Update);
+            pWriter.WriteInt(achieve.Id);
+            pWriter.Write(WriteIndividualAchieve(achieve).Buffer);
+
+            return pWriter;
+        }
+
+        private static Packet WriteIndividualAchieve(Achieve achieve)
+        {
+            GradeStatus mode = achieve.Condition == 0 ? GradeStatus.FinalGrade : GradeStatus.NotFinalGrade;
```

Comment:
This should be done outside this method and passed in as a parameter

Reply:
I can do that but then I will have to repeat that code in `WriteTableContent()` and `WriteUpdate()` since they both call that function. Is that fine?